### PR TITLE
Update PSE 8 with new question flow

### PR DESCRIPTION
**Pull request by kelsey-steven-ada**
Asana Tickets
- [Add AI Logical Steps Review to Unit 3 PSEs](https://app.asana.com/1/181459410160484/project/1205655776549324/task/1210177798860205?focus=true)
- [Update PSE 08 - Logical Steps](https://app.asana.com/1/181459410160484/project/1205655776549324/task/1202307845657322?focus=true)

Other Changes
- Adds example solution as an explanation to the implementation challenge
- Removes single letter variable names, which includes renaming the function since the function name references a single letter parameter
- Renames tests where relevant for better naming practices

### CODE REVIEW

File: 08/kth-missing-positive-number.md [hunk 1]
```diff
@@ -179,35 +295,94 @@ def test_empty_lists():
...
-4. Inside the loop:
-    1. Check if there are numbers missing between the current element and the next element in `arr`. Store the difference between the next element and the current element of `arr` in a new variable `missing_count`.
-    2. if `k` is less than or equal to `missing_count`, return the value of `arr` at `index` plus `k` 
-    3. Update the variable `k` by subtracting `missing_count`
-5. If the loop ends, return the last element of the list plus the remaining value of `k`.
+1. Handle edge case:
+   * If `missing_target` is smaller than the first value in `numbers`, return `missing_target`
+2. Create a loop over the input `numbers` that: 
+   * starts at the first element of `numbers` 
+   * ends one element before the last element of `numbers`
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]:
```suggestion
2. Loop over the input `numbers`: 
   * start at the first element of `numbers` 
   * end one element before the last element of `numbers`
```

File: 08/kth-missing-positive-number.md [hunk 2]
```diff
@@ -179,35 +295,94 @@ def test_empty_lists():
...
+3. Inside the loop:
+    1. Check if there are numbers missing between the current element and the next element in `numbers`. If there are missing numbers, store the count in a variable.
+    2. If `missing_target` is less than or equal to the count variable just created, return the value of `numbers` at the current index plus `missing_target`
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]:
```suggestion
    2. If `missing_target` is less than or equal to the gap size variable just created, the missing number is found within this gap. `missing_target` has been reduced with each gap we've found, so what's left is the number of values beyond the value at the current index. Return the value of `numbers` at the current index plus `missing_target` (the remaining count of missing values).
```

File: 08/kth-missing-positive-number.md [hunk 3]
```diff
@@ -179,35 +295,94 @@ def test_empty_lists():
...
+1. Handle edge case:
+   * If `missing_target` is smaller than the first value in `numbers`, return `missing_target`
+2. Create a loop over the input `numbers` that: 
+   * starts at the first element of `numbers` 
+   * ends one element before the last element of `numbers`
+3. Inside the loop:
+    1. Check if there are numbers missing between the current element and the next element in `numbers`. If there are missing numbers, store the count in a variable.
+    2. If `missing_target` is less than or equal to the count variable just created, return the value of `numbers` at the current index plus `missing_target` 
+    3. If we did not return, update the variable `missing_target` by subtracting the new missing count variable
+4. If the loop ends, return the last element of the list plus the remaining value of `missing_target`.
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]:
```suggestion
4. If the loop ends, this means that even accounting for all of the gaps in the array, there are still numbers left to find. `missing_target` has been reduced by the total size of the gaps found in the array, so the missing value we're looking for is the final value in the array plus the remaining count of missing values `missing_target`. Return the last element of the list plus the remaining value of `missing_target`.
```

File: 08/kth-missing-positive-number.md [hunk 4]
```diff
@@ -179,35 +295,94 @@ def test_empty_lists():
...
-2. If `k` is smaller than the first value in the input list `arr`, return `k`
-3. Create a loop over the input `arr` that: 
-   * starts at the first element of `arr` 
-   * ends one element before the last element of `arr`
-4. Inside the loop:
-    1. Check if there are numbers missing between the current element and the next element in `arr`. Store the difference between the next element and the current element of `arr` in a new variable `missing_count`.
-    2. if `k` is less than or equal to `missing_count`, return the value of `arr` at `index` plus `k` 
-    3. Update the variable `k` by subtracting `missing_count`
-5. If the loop ends, return the last element of the list plus the remaining value of `k`.
+1. Handle edge case:
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]: The extra content I added below is probably a bit wordy and could still be reworked, but what I'd like to see is more of a description of why all of this works. The meaning of the adjustment to `missing_target` feels non-obvious to me, so most of what I added is around making sense of that. But overall, for most of the places where we have a step like "if this, do that", I'd like it to be more of the form "if this, that means the following, therefore we do that."

File: 08/kth-missing-positive-number.md [hunk 5]
```diff
@@ -1,79 +1,195 @@
...
+- Did the AI tool uncover anything about the questions you asked that you hadn’t considered?
+- Did the AI tool suggest updates to the observations you made or questions you asked? 
+    - If so, what updates and why?
+- Did the AI tool suggest any new observations or questions?
+    - If so, what? Why would they be useful?
+
+##### !end-hint
+##### !explanation 
+
+For an example of what a review response might look like, let’s say that we used the example response from the "Explanation" section of the previous question to complete the review prompt.
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]: Similar update from PSE 7 to give us a little more wiggle room.
```suggestion
For an example of what a review response might look like, let’s say that we provided observations similar to the example response from the "Explanation" section of the previous question to complete the review prompt. 
```

**Review by anselrognlie — CHANGES_REQUESTED** [review]
Just wanted to list out the main themes in my feedback for visibility:
- The rename
  - `missing_target` doesn't sit that well with me, so I'm wondering whether we can go back to Kth Missing and use that `kth_missing` as the parameter name rather than just `k`.
- Missing tests
  - Due to an unindentation, Learn is only seeing 1 of the 4 tests
- Suggestions for observations/clarifications
  - Includes that small phrasing update from PSE 7 about how we got the sample feedback
- Suggestions for logical steps
  - Includes that small phrasing update from PSE 7 about how we got the sample

File: 08/kth-missing-positive-number.instructor.md [hunk 6]
```diff
@@ -1,53 +1,115 @@
...
-        # missing between arr[i] and arr[i + 1]
-        curr_missing = arr[i + 1] - arr[i] - 1
-        # if the kth missing is between
-        # arr[i] and arr[i + 1] -> return it
-        if k <= curr_missing:
-            return arr[i] + k
+def find_missing_positive_number(numbers, kth_missing):
+    # if the kth_missing is less than numbers[0]
+    if kth_missing <= numbers[0] - 1:
+        return kth_missing
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.instructor.md]: Update that can handle empty list
```suggestion
    # if numbers is empty, the kth missing value is kth_missing itself
    if not numbers:
        return kth_missing

    # if the kth_missing is less than numbers[0]
    if kth_missing <= numbers[0] - 1:
        return kth_missing
```

File: 08/kth-missing-positive-number.md [hunk 7]
```diff
@@ -86,7 +201,7 @@ Here are some example clarifying questions:
     * Consider at least one nominal and one edge case.
     * What is the expected output for the given input?
     * You can use the examples provided in the prompt, or other examples.
-2. Write unit tests for `kth_missing_positive_number` for the nominal and edge cases you identified in the first step.
+2. Write unit tests for `find_missing_positive_number` for the nominal and edge cases you identified in the first step.
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]: Regex replacement of "find_missing_positive_number" to "find_kth_missing_positive_number" worked fine in this file.

File: 08/kth-missing-positive-number.md [hunk 8]
```diff
@@ -179,35 +294,98 @@ def test_empty_lists():
...
-3. Create a loop over the input `arr` that: 
-   * starts at the first element of `arr` 
-   * ends one element before the last element of `arr`
-4. Inside the loop:
-    1. Check if there are numbers missing between the current element and the next element in `arr`. Store the difference between the next element and the current element of `arr` in a new variable `missing_count`.
-    2. if `k` is less than or equal to `missing_count`, return the value of `arr` at `index` plus `k` 
-    3. Update the variable `k` by subtracting `missing_count`
-5. If the loop ends, return the last element of the list plus the remaining value of `k`.
+1. Handle edge case:
+    * If `kth_missing` is smaller than the first value in `numbers`, then `kth_missing` is itself the missing number. Return `kth_missing`.
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]: Edge case for empty list, which out current linear time implementation would fail for.
```suggestion
    * If `numbers` is empty, then `kth_missing` is itself the missing number. Return `kth_missing`.
        * Since every number is missing, and the numbers and `kth_missing` both start from `1`, `kth_missing` is the missing number.
    * If `kth_missing` is smaller than the first value in `numbers`, then `kth_missing` is itself the missing number. Return `kth_missing`.
```

File: 08/kth-missing-positive-number.md [hunk 9]
```diff
@@ -179,35 +294,98 @@ def test_empty_lists():
...
+        * This is becasue it would take more numbers than `kth_missing` to even make us look at the array values.
```
Comment (anselrognlie) [inline 08/kth-missing-positive-number.md]:
```suggestion
        * This is because it would take more numbers than `kth_missing` to even make us look at the array values. 
```

**Review by anselrognlie — CHANGES_REQUESTED** [review]
Summary on this round
- Update statement a little to avoid mentioning position, and explained the meaning of kth
- Looked for any awkward wording due to the `kth_missing` variable renaming
  - Led to a reshuffling of the explanation obervations
- Noticed that the sample implementation wouldn't handle an empty list
  - Updated example implementation 9and instructtor copy)
  - Updated logical steps
- Restored kth as a part of the function name and tests
  - Just left comments about the regex I used and on specific exceptions, since that would be a ton of boilerplate comments.

**Review by anselrognlie — APPROVED** [review]
LGTM. Just one last typo.
